widgets/treeview/treeview.go

```diff
@@ -0,0 +1,818 @@
+package treeview
+
+import (
+	"errors"
+	"fmt"
+	"image"
+	"io"
+	"log"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/mum4k/termdash/cell"
+	"github.com/mum4k/termdash/keyboard"
+	"github.com/mum4k/termdash/mouse"
+	"github.com/mum4k/termdash/private/canvas"
+	"github.com/mum4k/termdash/private/runewidth"
+	"github.com/mum4k/termdash/terminal/terminalapi"
+	"github.com/mum4k/termdash/widgetapi"
+)
+
+// Number of nodes to scroll per mouse wheel event
+const (
+	ScrollStep = 5
+)
+
+// TreeNode represents a node in the treeview.
+type TreeNode struct {
+	ID            string
+	Label         string
+	Level         int
+	Parent        *TreeNode
+	Children      []*TreeNode
+	Value         interface{} // Can hold any data type
+	ShowSpinner   bool
+	OnClick       func() error
+	ExpandedState bool // Unique expanded state for each node
+	SpinnerIndex  int  // Current index for spinner icons
+	mu            sync.Mutex
+}
+
+// SetShowSpinner safely sets the ShowSpinner flag.
+func (node *TreeNode) SetShowSpinner(value bool) {
+	node.mu.Lock()
+	node.ShowSpinner = value
+	if !value {
+		node.SpinnerIndex = 0 // Reset spinner index when spinner is turned off
+	}
+	node.mu.Unlock()
+}
+
+// GetShowSpinner safely retrieves the ShowSpinner flag.
+func (node *TreeNode) GetShowSpinner() bool {
+	return node.ShowSpinner
```

Comment:
Since `node.ShowSpinner` is written under a mutex, we should also read it under a mutex for a complete thread safety.